Src/HALAL/Services/Communication/I2C/I2C.cpp

```diff
@@ -0,0 +1,248 @@
+    }
+
+    i2c->receive_ready = false;
+    return true;
+}
+
+void HAL_I2C_MasterRxCpltCallback(I2C_HandleTypeDef * hi2c){
+    auto result = find_if(I2C::active_i2c.begin(), I2C::active_i2c.end(), [&](auto i2c){return (i2c.second->hi2c == hi2c);});
+
+    if (result != I2C::active_i2c.end()) {
+        (*result).second->receive_ready = true;
+    }else{
+        //TODO: Warning: Data receive form an unknown SPI
+    }
+}
+
+
+bool I2C::read_from(uint8_t id, I2CPacket& packet, uint16_t mem_addr, uint16_t mem_size){
+	 if (!I2C::active_i2c.contains(id))
+	        return false; //TODO: Error handler
+
+	 I2C::Instance* i2c = I2C::active_i2c[id];
+
+
+	 *packet.get_data() = 0;
+
+	 if(i2c->hi2c->State == HAL_I2C_STATE_BUSY_TX){
+		printf("I2C Transmit buffer busy!\n\r");
+		return false;
+	 }
+
+	 if (HAL_I2C_Mem_Read_DMA(i2c->hi2c, packet.get_id(), mem_addr, 1, packet.get_data(), packet.get_size())){
+		 printf("I2C Error during receive!\n\r");
+		 return false; //TODO: Warning, Error during receive
+	 }
+
+	 i2c->receive_ready = false;
+	 return true;
+}
+
+bool I2C::write_to(uint8_t id, I2CPacket& packet, uint16_t mem_addr, uint16_t mem_size){
+	if (!I2C::active_i2c.contains(id))
+		return false; //TODO: Error handler
+
+	I2C::Instance* i2c = I2C::active_i2c[id];
+
+	 if(i2c->hi2c->State == HAL_I2C_STATE_BUSY_RX){
+		printf("I2C Receive buffer busy!\n\r");
+	   return false;
+	 }
+
+	if (HAL_I2C_Mem_Write_DMA(i2c->hi2c, packet.get_id(), mem_addr, mem_size, packet.get_data(), packet.get_size())){
+		return false; //TODO: Warning, Error during transmision
+	}
+
+	return true;
+
+}
+
+bool I2C::has_next_packet(uint8_t id){
+    if (!I2C::active_i2c.contains(id))
+        return false; //TODO: Error handler
+
+    I2C::Instance* i2c = I2C::active_i2c[id];
+
+    return i2c->receive_ready;
+}
+
+bool I2C::is_busy(uint8_t id){
+    if (!I2C::active_i2c.contains(id))
+        return false; //TODO: Error handler
+
+    I2C::Instance* i2c = I2C::active_i2c[id];
+
+    if(i2c->hi2c->State == HAL_I2C_STATE_BUSY_TX){
+    	return true;
+    }
+
+
+    return false;
+}
+
+void HAL_I2C_ErrorCallback(I2C_HandleTypeDef *hspi){
+    //TODO: Fault, SPI error
+    return;
+}
+
+
+void I2C::init(I2C::Instance* i2c){
+	if(i2c->initialized){
+		return;
+	}
+	i2c->hi2c->Instance = i2c->instance;
+	i2c->hi2c->Init.Timing = 0x60404E72;
+	i2c->hi2c->Init.OwnAddress1 = (i2c->address)<<1;
+	i2c->hi2c->Init.AddressingMode = I2C_ADDRESSINGMODE_7BIT;
+	i2c->hi2c->Init.DualAddressMode = I2C_DUALADDRESS_DISABLE;
+	i2c->hi2c->Init.OwnAddress2 = 0;
+	i2c->hi2c->Init.OwnAddress2Masks = I2C_OA2_NOMASK;
+	i2c->hi2c->Init.GeneralCallMode = I2C_GENERALCALL_DISABLE;
+	i2c->hi2c->Init.NoStretchMode = I2C_NOSTRETCH_DISABLE;
+
+	if (HAL_I2C_Init(i2c->hi2c) != HAL_OK){
+		ErrorHandler("Error configurating I2C");
+	  }
+
+	if (HAL_I2CEx_ConfigAnalogFilter(i2c->hi2c, I2C_ANALOGFILTER_ENABLE) != HAL_OK){
+		ErrorHandler("Error configurating Analog Filter of the I2C");
```

Comment:
```suggestion
		ErrorHandler("Error configurating Analog Filter of the I2C");
```
Lo mismo que los errores de arriba. Cada error ha de ser concreto, informativo y unico